app/src/main/java/org/schabi/newpipe/player/Player.java

```diff
@@ -3262,11 +3306,83 @@ public void disablePreloadingOfCurrentTrack() {
     }
 
+    /**
+     * Set the default resolution for livestreams according to user choice in "Video and audio"
+     * settings.
+     *
+     * <p>
+     * If no stream match the minimum criteria required (video format height <= default height),
+     * then the maximum resolution limit and the maximum frame rate will be set.
+     * </p>
+     *
+     * <p>
+     * Note that the best way should be to create a custom
+     * {@link com.google.android.exoplayer2.trackselection.AdaptiveTrackSelection} instead
+     * because it will never load the automatically selected stream but that's a bit more complex
+     * to implement this.
+     * </p>
+     *
+     * @param videoTrackGroupArray the video {@link TrackGroupArray}, which cannot be null
+     * @param videoRendererIndex   the video renderer index
+     */
+    private void setDefaultResolutionForLivestreamsIfNeeded(
+            @NonNull final TrackGroupArray videoTrackGroupArray,
+            final int videoRendererIndex) {
+        if (isDefaultResolutionSetForLivestreams) {
+            return;
+        }
+
+        final String defaultResolution;
+        if (popupPlayerSelected()) {
+            defaultResolution = computeDefaultResolution(context,
+                    R.string.default_popup_resolution_key,
+                    R.string.default_popup_resolution_value);
+        } else {
+            defaultResolution = computeDefaultResolution(context,
+                    R.string.default_resolution_key, R.string.default_resolution_value);
+        }
+        final String[] defaultResolutionArray = defaultResolution.split("p");
+        final int defaultHeight = !defaultResolution.equals(context.getString(
+                R.string.best_resolution_key))
+                ? Integer.parseInt(defaultResolutionArray[0]) : -1;
+        final int defaultFrameRate = defaultResolutionArray.length == 2
+                ? Integer.parseInt(defaultResolutionArray[1]) : -1;
+
+        final DefaultTrackSelector.Parameters selectionParameters = trackSelector.getParameters();
+        final DefaultTrackSelector.ParametersBuilder builder = selectionParameters.buildUpon()
+                .clearSelectionOverrides();
+
+        for (int i = 0; i < videoTrackGroupArray.length; i++) {
+            final TrackGroup trackGroup = videoTrackGroupArray.get(i);
+            for (int j = trackGroup.length - 1; j >= 0; j--) {
+                final Format videoFormat = trackGroup.getFormat(j);
+                if ((defaultHeight == -1 || videoFormat.height <= defaultHeight)
+                        && (defaultFrameRate <= -1 || videoFormat.frameRate <= defaultFrameRate)) {
+                    builder.setSelectionOverride(videoRendererIndex, videoTrackGroupArray,
+                            new DefaultTrackSelector.SelectionOverride(videoRendererIndex, j));
+                    isDefaultResolutionSetForLivestreams = true;
+                    break;
+                }
+            }
+        }
```

Comment:
From what I can understand, this `for` overrides the selection for each possible `trackGroup` on the first possible `videoFormat` that satisfies `(defaultHeight == -1 || videoFormat.height <= defaultHeight) && (defaultFrameRate <= -1 || videoFormat.frameRate <= defaultFrameRate)`. A couple of questions:
- why `defaultFrameRate <= -1` and not `defaultFrameRate == -1` for consistency?
- why stop at the first suitable track? Maybe there are more similar tracks next. Wouldn't this always choose 144p?
- (question about exoplayer) why are there multiple formats for each track? Shouldn't each track have exactly one format? Otherwise it would be split into multiple tracks... I don't understand exoplayer...

Reply:
Any news @TiA4f8R ?